src/data_structures/bit_tree.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright 2014-2015 Johannes Köster, Vadim Nazarov, Patrick Marks
+// Licensed under the MIT license (http://opensource.org/licenses/MIT)
+// This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! BIT-tree (Binary Indexed Trees, aka Fenwick Tree) maintains a prefix-sum or
+//! prefix-max that can be efficiently queried and updated. From: Peter M. Fenwick (1994). "A new data structure for cumulative frequency tables". Software: Practice and Experience. 24 (3): 327–336. 
+//! Implementation outlined here: https://www.topcoder.com/community/data-science/data-science-tutorials/binary-indexed-trees/
+//!
+//! # Example
+//!
+//! ```
+//! use bio::data_structures::bit_tree::*;
+//!
+//! let mut bit = MaxBitTree::new(10);
+//! bit.set(0, (1,0));
+//! bit.set(1, (0,1));
+//! bit.set(2, (2,2));
+//! bit.set(3, (4,3));
+//!
+//! assert_eq!(bit.get(0), (1, 0));
+//! assert_eq!(bit.get(1), (1, 0));
+//! assert_eq!(bit.get(2), (2, 2));
+//! assert_eq!(bit.get(3), (4, 3));
+//! assert_eq!(bit.get(4), (4, 3));
+
+use std::cmp::max;
+
+/// Max Bit Tree. get(i) will return the largest element e that has been added
+/// to the bit tree with set(j, e), where j <= i. Initially all positions have
+/// the value T::default(). Note that a set cannot be 'undone' by inserting
+/// a smaller element at the same index.
+pub struct MaxBitTree<T: Default + Ord> {
+    tree: Vec<T>,
+}
+
+impl<T: Ord + Default + Copy> MaxBitTree<T> {
+    /// Create a new bit tree with len elements
+    pub fn new(len: usize) -> MaxBitTree<T> {
+        let mut tree = Vec::new();
+
+        // Pad length by one. The first element is unused.  
+        // Done this way to make the tree structure work correctly.
+        for _ in 0..(len+2) {
+            tree.push(T::default());
+        }
+
+        MaxBitTree { tree: tree }
+    }
+
+    /// get(i) returns the largest element e that has been added
+    /// to the bit tree with set(j, e), where j <= i. 
+    pub fn get(&self, idx: usize) -> T {
+        let mut idx = idx + 1;
+        let mut sum = T::default();
+        while idx > 0 {
+            sum = max(sum, self.tree[idx].clone());
```

Comment:
I guess the code would be the same for a prefix-sum? Can this be generalized into a trait `FenwickTree` trait, and you derive `MaxFenwickTree` and `SumFenwickTree`, only implementing a method `operation(T, T)`?